File: backend/app/shor.py
"""Classical period finding and factoring driving the Shor visualization."""
from __future__ import annotations

import math
import random
from typing import Optional


def find_period(a: int, N: int) -> int:
    if math.gcd(a, N) != 1:
        return 0
    value = a % N
    r = 1
    while value != 1:
        value = (value * a) % N
        r += 1
        if r > N:
            return 0
    return r


def factor(N: int, a: Optional[int] = None) -> dict:
    steps: list[str] = []
    if N % 2 == 0:
        return {"N": N, "a": 2, "period": 1, "factors": [2, N // 2], "steps": ["N is even."]}

    if a is None:
        candidates = [x for x in range(2, N) if math.gcd(x, N) == 1]
        a = random.choice(candidates) if candidates else 2

    g = math.gcd(a, N)
    if g != 1:
        return {"N": N, "a": a, "period": 0, "factors": [g, N // g], "steps": [f"gcd({a},{N})={g} - lucky guess."]}

    r = find_period(a, N)
    steps.append(f"Chose a={a}, coprime to N={N}.")
    steps.append(f"Period r={r} (quantum order finding): {a}^{r} ≡ 1 (mod {N}).")

    factors: Optional[list[int]] = None
    if r and r % 2 == 0:
        half = pow(a, r // 2, N)
        if half != N - 1:
            f1, f2 = math.gcd(half - 1, N), math.gcd(half + 1, N)
            steps.append(f"gcd({a}^(r/2)±1, N) = gcd({half-1},{N}), gcd({half+1},{N}).")
            for f in (f1, f2):
                if 1 < f < N:
                    factors = [f, N // f]
                    break
        else:
            steps.append("a^(r/2) ≡ -1 (mod N) - failure, retry.")
    else:
        steps.append("Period is odd - retry with a different a.")

    if factors:
        steps.append(f"Factors of {N}: {factors[0]} × {factors[1]}.")
    return {"N": N, "a": a, "period": r, "factors": factors, "steps": steps}

File: frontend/src/quantum/shor.ts
/**
 * Classical period-finding + factoring helpers used to drive the Shor
 * visualization. The quantum subroutine (order finding) is illustrated; the
 * surrounding number theory runs classically here for didactic clarity.
 */

export function gcd(a: number, b: number): number {
  while (b) {
    [a, b] = [b, a % b];
  }
  return Math.abs(a);
}

export function modPow(base: number, exp: number, mod: number): number {
  let result = 1;
  base %= mod;
  while (exp > 0) {
    if (exp & 1) result = (result * base) % mod;
    exp >>= 1;
    base = (base * base) % mod;
  }
  return result;
}

/** Find the multiplicative order r of a mod N (smallest r with a^r ≡ 1). */
export function findPeriod(a: number, N: number): number {
  if (gcd(a, N) !== 1) return 0;
  let r = 1;
  let value = a % N;
  while (value !== 1) {
    value = (value * a) % N;
    r++;
    if (r > N) return 0;
  }
  return r;
}

export interface ShorResult {
  N: number;
  a: number;
  period: number;
  factors: [number, number] | null;
  sequence: number[];
  steps: string[];
}

/** Run the Shor pipeline for a chosen a, returning everything to visualize. */
export function runShor(N: number, a: number): ShorResult {
  const steps: string[] = [];
  const g = gcd(a, N);
  if (g !== 1) {
    return {
      N,
      a,
      period: 0,
      factors: [g, N / g],
      sequence: [],
      steps: [`gcd(${a}, ${N}) = ${g} ≠ 1 - lucky guess gives factors directly.`],
    };
  }
  const period = findPeriod(a, N);
  const sequence: number[] = [];
  for (let x = 0; x < Math.min(period + 1, 32); x++) sequence.push(modPow(a, x, N));
  steps.push(`Chose a = ${a}, coprime to N = ${N}.`);
  steps.push(`Period finding (quantum) gives r = ${period} since ${a}^${period} ≡ 1 (mod ${N}).`);

  let factors: [number, number] | null = null;
  if (period % 2 === 0) {
    const half = modPow(a, period / 2, N);
    if (half !== N - 1) {
      const f1 = gcd(half - 1, N);
      const f2 = gcd(half + 1, N);
      steps.push(`Compute gcd(${a}^{r/2} ± 1, N) = gcd(${half - 1}, ${N}) and gcd(${half + 1}, ${N}).`);
      if (f1 > 1 && f1 < N) factors = [f1, N / f1];
      else if (f2 > 1 && f2 < N) factors = [f2, N / f2];
    } else {
      steps.push(`a^{r/2} ≡ -1 (mod N) - failure, retry with a different a.`);
    }
  } else {
    steps.push(`Period r = ${period} is odd - retry with a different a.`);
  }
  if (factors) steps.push(`Factors of ${N}: ${factors[0]} × ${factors[1]}.`);
  return { N, a, period, factors, sequence, steps };
}

/** Candidate a values coprime to N, for the UI selector. */
export function coprimes(N: number): number[] {
  const out: number[] = [];
  for (let a = 2; a < N; a++) if (gcd(a, N) === 1) out.push(a);
  return out;
}
